### DISCUSSION TITLE
Repeating audio stimuli after correct answer (feedback condition)

### DISCUSSION THREAD

**Discussion by FelicityF** [jspsych/jsPsych#843] [Q&A] [answered]
For my experiment participants will be seeing four scenes at the screen while listening to a sentence from an artificial language. Their task is to select the scene that they think matches the sentences they just heard. I managed to program this part and the data is correctly recording as well. However, we have 3 feedback conditions that I have to implement and that I could use some help with. 

- Positive feedback: after correctly selecting the target scene the spoken sequence will be repeated while the foil scenes disappear, leaving only the correct picture on the screen. 
- Negative feedback: no form of explicit feedback will be given after selecting the correct scene. After selecting a foil, the correct spoken sequence and a red X will be presented. 
- No feedback: No feedback will be given. However, the auditory sequence will be repeated after each mouse click. 

**Question 1**

How could I repeat the auditory stimuli once after selecting a scene (no feedback condition), and how could I do this conditionally (positive feedback condition)?

**Question 2**

For the positive feedback condition, is it possible to let the three other scenes disappear or let the target scene lit up –emphasising their (correct) response?

**Question 3**

Within the positive and negative feedback conditions, there are sub conditions. That is, some participants will receive feedback 50% of the time, whereas others will receive feedback 75% of the time. Would there be a way to implement this in the code from question 1? 

Many thanks for your help and suggestions.

**Comment by jodeleeuw** [comment, marked as answer]
Hi @FelicityF,

> How could I repeat the auditory stimuli once after selecting a scene (no feedback condition), and how could I do this conditionally (positive feedback condition)?

I think this depends a bit on how you've set up the task so far. One option would be to repeat the trial again with some slightly different parameters. For example, if you are using the `audio-button-response` plugin to run these trials then you could show the trial again but use CSS to set the `visibility: hidden` property of all the foil responses. This would hide them, but keep the same structure on the screen so that the target doesn't move. 

To do it conditionally, you'll need to use a [conditional timeline](https://www.jspsych.org/overview/timeline/#conditional-timelines). You can check if the response was correct in the last trial and only show the feedback trial if it was.

> For the positive feedback condition, is it possible to let the three other scenes disappear or let the target scene lit up –emphasising their (correct) response?

Yes, I think my suggestion above is one way to do this. It would hide the other three. You could also use [CSS image filters](https://www.w3schools.com/howto/howto_css_image_effects.asp) or [CSS opacity](https://www.w3schools.com/css/css_image_transparency.asp) to de-emphasize the foils relative to the target. 

> Within the positive and negative feedback conditions, there are sub conditions. That is, some participants will receive feedback 50% of the time, whereas others will receive feedback 75% of the time. Would there be a way to implement this in the code from question 1?

You could implement this logic inside the `conditional_function` for the conditional timeline. For example:

```js
var condition_feedback_probability = 0.75; // let's assume they are in the 75% condition.

var feedback = {
	timeline: [ feedback_trial ],
	conditional_function: function() {
		// let's assume they are in the positive feedback condition 
		// and we show feedback if there is a correct response
		var show_feedback = jsPsych.data.get().last(1).values()[0].correct; // true if last response is correct.
		if(show_feedback){
			var random_number = Math.random(); // uniform distribution between 0-1
			if(random_number < condition_feedback_probability){
				return true;
			}
		}
		return false;
	}
}
```

**Reply by FelicityF** [reply]
Hi @jodeleeuw 

Thank you for your reply. This is the code I'm having so far:

```js
  var fixation = {
        type: 'html-keyboard-response',
        stimulus: '<span style="font-size:40px;">+</span>',
        choices: jsPsych.NO_KEYS,
        trial_duration: 1000
    }


    var PGG_Positive = { 
       type: 'audio-button-response', 
       stimulus: jsPsych.timelineVariable('audio'), 
       choices: [jsPsych.timelineVariable('Target'),
              jsPsych.timelineVariable('Foil1'),
              jsPsych.timelineVariable('Foil2'),
              jsPsych.timelineVariable('Foil3')],
       button_html:'<img style = "width:100px;" src = "%choice%" />',
       margin_horizontal: "20px",
       data: jsPsych.timelineVariable('data'),
       on_finish: function(data){
       if(data.button_pressed == '0'){
        data.response = 'Target';
      } else if (data.button_pressed == '1'){
        data.response = 'Foil1';
      } else if (data.button_pressed == '2'){
        data.response = 'Foil2';
      } else {
        data.response = 'Foil3';
      }
      if(data.button_pressed == data.button){
        data.response_correct = true;
      } else {
        data.response_correct = false;
      }
    }
   }

    var procedure = { 
        timeline: [fixation, PGG_Positive], 
        timeline_variables: [ 
        {audio: 'sound/hef hep jove poox.mp3', data: {button: '0', focus: 'Target'}, Target: 'img/PGG.png', Foil1: 
         'img/blue.png', Foil2: 'img/orange.png', Foil3: 'img/redX.png'},
        {audio: 'sound/hef hep jove poox.mp3', data: {button: '1', focus: 'Foil1'}, Target: 'img/PGG.png', Foil1: 
        'img/blue.png', Foil2: 'img/blue.png', Foil3: 'img/blue.png'},
    ]}

    timeline.push(procedure);
```

I'm not entirely sure where to implement the suggestions you posted earlier, so if anyone has some advice that would be extremely helpful. 

Thank you!

**Reply by FelicityF** [reply]
@jodeleeuw I changed the code slightly and I got it to work! When clicking on a foil, the next trial will be presented without auditory feedback, but when clicking on the target, the auditory stimuli is repeated (just as I wanted). However, after the auditory sequence is repeated it is required to click again before proceeding to the next trial (I assume this is because I'm using timeline: [ PGG_Positive ]). Is there a way to avoid this? Thus: correct > auditory sequence repeated > next trial rather than correct > auditory sequence repeated > respond again > next trial.

**Reply by jodeleeuw** [reply]
Great! What kind of trial are you using to repeat the auditory stimulus? For many plugins you can set the `trial_duration` parameter to automatically end the trial.

**Reply by FelicityF** [reply]
This is the code that I'm currently using: 

```js

var PGG_Positive = {
        type: 'audio-button-response',
        stimulus: jsPsych.timelineVariable('audio'),
        choices: jsPsych.timelineVariable('choices'),
        button_html:'<img style = "width:400px;" src = "%choice%" />',
        margin_horizontal: "60px",
        margin_vertical: "60px",
        data: jsPsych.timelineVariable('choices'),
        on_finish: function(data){
          if(data.button_pressed == '0'){
          data.response = 'Target';
        } else if (data.button_pressed == '1'){
          data.response = 'Foil1';
        } else if (data.button_pressed == '2'){
          data.response = 'Foil2';
        } else {
          data.response = 'Foil3';
        }
          if(data.button_pressed == 0){
            data.response_correct = true;
          } else {
            data.response_correct = false;
          }
        }
      };

      var condition_feedback_probability = 0.75;

      var feedback = {
	       timeline: [ PGG_Positive ],
	        conditional_function: function() {
		  var show_feedback = jsPsych.data.get().last(1).values()[0].response_correct; 
		  if(show_feedback){
			var random_number = Math.random(); 
			if(random_number < condition_feedback_probability){
				return true;
			}
		}
		return false;
	}
}
```

I could add ```trial_duration:4000``` after ```type: 'audio-button-response'``` to make sure that the next trail is presented after 4s. However, this also affects the time that participants are able to make their initial response obviously. Although this seems a reasonable solution, I'm wondering whether there is a way to work around that so this time constrain is only applied within:

``` js
var feedback = {
	       timeline: [ PGG_Positive ],
	        conditional_function: function() {
		  var show_feedback = jsPsych.data.get().last(1).values()[0].response_correct; 
		  if(show_feedback){
			var random_number = Math.random(); 
			if(random_number < condition_feedback_probability){
				return true;
			}
		}
		return false;
	}
}
```

Additionally, I haven't figured out yet how to hide the foils after selecting the target. Because we only want to hide the foils after their response, I thought this code should possibly also be implemented in the feedback part of the code? Any suggestions here? I really appreciate all the help!

**Reply by jodeleeuw** [reply]
I think you'll want to make a copy of the `PGG_Positive` trial so that you can modify the parameters separately from the main trial. You can still use the same `stimulus` and `choices` parameters in the copy so that the display matches, but you could set `trial_duration` in the second as well as perhaps `response_ends_trial: false` to ensure that any response is ignored.

For hiding the foils, you may need to pass an array to the `button_html` property. This will let you style each button separately. For example, this would hide the 2nd, 3rd, and 4th items.

```js
button_html: [
	'<img style = "width:400px;" src = "%choice%" />',
	'<img style = "width:400px; visibility: hidden;" src = "%choice%" />',
	'<img style = "width:400px; visibility: hidden;" src = "%choice%" />',
	'<img style = "width:400px; visibility: hidden;" src = "%choice%" />'
]
```

One tricky part of this will be to hide the right items, since the target won't be in the same location each time. For this, you may want to add a timeline variable that is the button_html array so that the correct array is linked with each trial.

**Reply by FelicityF** [reply]
Ah yes, that works. Thank you so much! The correct items are indeed hidden for certain trials, so I will try to figure out how to create and implement the button_html array. After that it should work!
